internal/provider/resource_gitlab_project_issue.go

```diff
@@ -0,0 +1,294 @@
+	}
+
+	options := &gitlab.UpdateIssueOptions{}
+	if d.HasChange("title") {
+		options.Title = gitlab.String(d.Get("title").(string))
+	}
+	if d.HasChange("assignee_ids") {
+		options.AssigneeIDs = intSetToIntSlice(d.Get("assignee_ids").(*schema.Set))
+	}
+	if d.HasChange("confidential") {
+		options.Confidential = gitlab.Bool(d.Get("confidential").(bool))
+	}
+	if d.HasChange("description") {
+		options.Description = gitlab.String(d.Get("description").(string))
+	}
+	if d.HasChange("due_date") {
+		dueDate := d.Get("due_date").(string)
+		if dueDate != "" {
+			parsedDueDate, err := parseISO8601Date(dueDate)
+			if err != nil {
+				return diag.Errorf("failed to parse due_date: %s. %v", dueDate, err)
+			}
+			options.DueDate = parsedDueDate
+		} else {
+			// see https://github.com/xanzy/go-gitlab/issues/1384
+			return diag.Errorf("remove a due date is currently not supported. See https://github.com/xanzy/go-gitlab/issues/1384")
+		}
+	}
+	if d.HasChange("issue_type") {
+		options.IssueType = gitlab.String(d.Get("issue_type").(string))
+	}
+	if d.HasChange("labels") {
+		gitlabLabels := gitlab.Labels(*stringSetToStringSlice(d.Get("labels").(*schema.Set)))
+		options.Labels = &gitlabLabels
+	}
+	if d.HasChange("milestone_id") {
+		options.MilestoneID = gitlab.Int(d.Get("milestone_id").(int))
+	}
+	if d.HasChange("weight") {
+		options.Weight = gitlab.Int(d.Get("weight").(int))
+	}
+	if d.HasChange("state") {
+		options.StateEvent = gitlab.String(issueStateToStateEvent[d.Get("state").(string)])
+	}
+
+	_, _, err = client.Issues.UpdateIssue(project, issueIID, options, gitlab.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceGitlabProjectIssueRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectIssueDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	project, issueIID, err := resourceGitLabProjectIssueParseId(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	deleteOnDestroy := d.Get("delete_on_destroy").(bool)
+
+	if deleteOnDestroy {
+		log.Printf("[DEBUG] Deleting issue %d in project %s for destroy", issueIID, project)
+		resp, err := client.Issues.DeleteIssue(project, issueIID, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.Errorf("%s failed to delete issue %d in project %s: (%s) %v", d.Id(), issueIID, project, resp.Status, err)
+		}
+	} else {
+		log.Printf("[DEBUG] Closing issue %d in project %s for destroy", issueIID, project)
+		_, resp, err := client.Issues.UpdateIssue(project, issueIID, &gitlab.UpdateIssueOptions{StateEvent: gitlab.String("close")}, gitlab.WithContext(ctx))
+		if err != nil {
+			return diag.Errorf("%s failed to delete issue %d in project %s: (%s) %v", d.Id(), issueIID, project, resp.Status, err)
+		}
+	}
+
+	return nil
+}
+
+func resourceGitLabProjectIssueParseId(id string) (string, int, error) {
+	project, issue, err := parseTwoPartID(id)
+	if err != nil {
+		return "", -1, err
```

Comment:
Go convention is to use the zero-value when there is an error. The caller is expected to inspect the error first.